### ISSUE TITLE
DeprecationWarning: 'pipes' is deprecated and slated for removal in Python 3.13

### ISSUE THREAD

**Issue by danielhollas** [circus-tent/circus#1215]
The following DeprecationWarning started appearing in Python 3.11 and will become relevant later this year when Python 3.13 is released

```python
    from circus.client import CircusClient
/opt/hostedtoolcache/Python/3.11.9/x64/lib/python3.11/site-packages/circus/client.py:11: in <module>
    from circus.util import DEFAULT_ENDPOINT_DEALER, get_connection, to_bytes
/opt/hostedtoolcache/Python/3.11.9/x64/lib/python3.11/site-packages/circus/util.py:42: in <module>
    from pipes import quote as shell_escape_arg
/opt/hostedtoolcache/Python/3.11.9/x64/lib/python3.11/pipes.py:68: in <module>
    warnings._deprecated(__name__, remove=(3, 13))
DeprecationWarning: 'pipes' is deprecated and slated for removal in Python 3.13
```

The [pipes module](https://docs.python.org/3/library/pipes.html#module-pipes) was deprecated in PEP-594:

> The [pipes](https://docs.python.org/3.12/library/pipes.html#module-pipes) module provides helpers to pipe the input of one command into the output of another command. The module is built on top of os.popen. Users are encouraged to use the [subprocess](https://docs.python.org/3/library/subprocess.html#module-subprocess) module instead.

https://peps.python.org/pep-0594/#pipes

**Comment by GeigerJ2** [comment]
Just stumbled on this issue, as well, while using Python 3.13, in which `pipes` is actually removed (leading to a `ModuleNotFoundError` when calling `verdi status` in AiiDA, originating from this library). I realized that this is fixed on main, [see here](https://github.com/circus-tent/circus/pull/1211) (@sphuber is everywhere hehe). Not sure how well-maintained the library is at this point, as the last changes on `master`, as well as the v0.18.0 release, are abut 2 years in the past.

**Comment by sphuber** [comment]
There has indeed been almost no maintenance of this package. De facto it has been me and @biozz most recently, but we are clearly not active. It seems that the PR you linked is the only real fix on the `master` branch since the v0.18.0 release. So I guess I could simply trigger a new patch release. I seem to have added a CD workflow to automate it, so it should be straight forward, but it has been a while :sweat_smile:

**Comment by blokhin** [comment]
the PyPI package should be updated, since `0.18.0` still imports `pipes`

**Comment by sphuber** [comment]
We are working on a 0.19.0 release which should fix it: https://github.com/circus-tent/circus/pull/1223
@agoscinski can you please update the PR to fix the change notes and we can release it

**Comment by sphuber** [comment]
The release commit was made and merged and I have made the tag, but I don't have the credentials for PyPI. I thought I had it as I seem to remember I release 0.18.0, but maybe I am misremembering or my access was removed. Anyway, I cannot push to PyPI at the moment.  @biozz @k4nar do you guys have access? If so, could you push the new release 0.19.0 to PyPI or give my account (https://pypi.org/user/sphuber/) access?

**Comment by k4nar** [comment]
@sphuber : you are supposed to be able to publish a release as a maintainer 🤔 .

<img width="826" alt="Image" src="https://github.com/user-attachments/assets/94684934-af12-420f-9762-58192e012e4c" />

**Comment by danielhollas** [comment]
Based on the CI failure, it looks like the PYPI_KEY secret is not set in the repo.

https://github.com/circus-tent/circus/blob/ce7bab244cdbc736591517fbddb4f92836906129/.github/workflows/cd.yml#L121C1-L121C55

**Comment by sphuber** [comment]
> Based on the CI failure, it looks like the PYPI_KEY secret is not set in the repo.

Yes, I noticed that, so I went to PyPI to generate a token but I cannot. Apparently maintainer doesn't give sufficient rights:

![Image](https://github.com/user-attachments/assets/3c52f9fc-1076-4bfe-8290-d814e02ef27f)


Not that it would matter, since I also don't have sufficient permissions to add that token here on Github as a secret to make the continuous deployment workflow work.

I guess I built and pushed to PyPI manually the last few releases (just couldn't remember that anymore). Will try to do that again tonight, don't have time now.

**Comment by k4nar** [comment]
I didn’t know (or remember) that we had CI steps to publish the releases. I’m not a huge fan of it as it exposes us to a leak of the token if we’re not very careful about the content of the PRs running the CI.

I don’t think that building & pushing the release locally is a huge burden given the very slow cadence of Circus releases :) .

**Comment by danielhollas** [comment]
> I didn’t know (or remember) that we had CI steps to publish the releases. I’m not a huge fan of it as it exposes us to a leak of the token if we’re not very careful about the content of the PRs running the CI.

If this is the concern, you can use the new [Trusted Publisher PyPI](https://docs.pypi.org/trusted-publishers/) functionality, which is tokenless and is designed exactly to alleviate this problem.  

You can find some more info/links here: https://github.com/aiidalab/aiidalab-widgets-base/issues/567

I've set it up on couple of my repos and the whole process was very smooth, here's an example Github workflow:

https://github.com/ispg-group/harmonwig/blob/main/.github/workflows/cd.yml

(where you can see that it doesn't include any tokens)

**Comment by sphuber** [comment]
> I don’t think that building & pushing the release locally is a huge burden given the very slow cadence of Circus releases :) .

Fair, and I just tried and do that but it seems that is no longer supported:

```
requests.exceptions.HTTPError: 403 Client Error: Username/Password authentication is no longer supported. Migrate to API Tokens or Trusted Publishers instead. See https://pypi.org/help/#apitoken and https://pypi.org/help/#trusted-publishers for url: https://upload.pypi.org/legacy/
```

So I think we are forced to use an API token or a Trusted Publisher, but I don't have the necessary rights to set up either.

**Comment by k4nar** [comment]
@danielhollas : wow I didn’t know about this, I haven’t published anything on PyPI since quite some time. That’s a great feature. If someones setups the CI for it, I can create the access on PyPI.

@sphuber : I think you should be able to create a personal API token here: https://pypi.org/manage/account/token/?selected_project=circus

**Comment by sphuber** [comment]
> [@sphuber](https://github.com/sphuber) : I think you should be able to create a personal API token here: https://pypi.org/manage/account/token/?selected_project=circus

Thanks @k4nar didn't know this. Just published 0.19.0: https://pypi.org/project/circus/#history
so this can now be closed
